gpytorch/lazy/diag_lazy_tensor.py

```diff
@@ -8,15 +8,18 @@
 
 
 class DiagLazyTensor(LazyTensor):
-    def __init__(self, diag):
+    def __init__(self, diag, exact_root_decomposition=False):
```

Comment:
Yeah - for large DiagLazyTensors this would use lots of memory

Reply:
Is there anywhere in gpytorch that we couldn't allow root decomposition to return another lazy tensor (specifically a `DiagLazyTensor` in this case obviously)? If we're using it for things like sampling and LOVE, it seems like we'd really only need `matmul` calls with it

Reply:
`MulLazyTensor` needs an exact tensor.

Reply:
For sampling, DiagLazyTensor already does the smart thing. For LOVE - you're right we maybe want to think of something smarter.

Reply:
Looking at `MulLazyTensor`, we only use the roots as tensors to accomplish elementwise multiplies. E.g. `right_root` in
https://github.com/cornellius-gp/gpytorch/blob/38f610c8b678fd38efd6d40379afc8293b49ca37/gpytorch/lazy/mul_lazy_tensor.py#L157

If we special case `DiagLazyTensor.__mul__` to:
1. Elementwise multiply the diagonal of `rhs` if `rhs` is a normal Tensor, and
2. Accomplish the diagonal elementwise multiplication with a diagonal correction if `rhs` is a LazyTensor

then I think we can fix all the call sites in `MulLazyTensor`. Another option would just be to special case in `MulLazyTensor` the scenario when `RootLazyTensor.root` is a `DiagLazyTensor`.

Reply:
Or `MulLazyTensor` could just explicitly toggle `exact_root_decomposition` to false if it is decomposing a `DiagLazyTensor` because let's be honest SKIP is always going to have some pretty gross special cases because of how ugly Hadamard products are.

Reply:
MulLazyTensor needs tensor roots for the derivative.

Reply:
We should maybe open a different issue for this? Right now root decomposition always returns a tensor, and so this would be have to be part of a much larger change.

Reply:
I feel we should definitely allow root decomposition to return a LazyTensor, this should help with memory footprint (an potentially wall time) in cases where DiagLazyTensors or BlockDiagLazyTensors are involved.

Reply:
@Balandat #354 does exactly this :-). I'm going to merge this in first, and then we can discuss that change to RootLazyTensor in that PR